File: problem931.py
class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]):
        # It is dp problem 
        # From bot to top

        for row in range(len(matrix)-1, 0, -1):
            # there are two options for next number 
            # first one is when it is two end of row
            for col in range(len(matrix)):
                if col == 0:
                    matrix[row-1][col] += min(matrix[row][col], matrix[row][col+1])
                elif col == len(matrix)-1:
                    matrix[row-1][col] += min(matrix[row][col-1], matrix[row][col])
                else:
                    matrix[row-1][col] += min(matrix[row][col-1], matrix[row][col], matrix[row][col+1])
        
        return min(matrix[0])

File: problem931.cpp
#include <algorithm>
#include <vector>
#include <iostream>
using namespace std;
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        // It is DP problem and it start from bottom to up

        int n = matrix.size();

        for(int row = n -2; row > -1; row--){
            for(int col = 0; col < n; col++ ){
                if(col == 0){
                    matrix[row][col] += min(matrix[row+1][col], matrix[row+1][col+1]);
                }
                else if (col == n-1){
                    matrix[row][col] += min(matrix[row+1][col], matrix[row+1][col-1]);
                }
                else{
                    int smallnum = min(matrix[row+1][col], matrix[row+1][col-1]);
                    matrix[row][col] += min(smallnum, matrix[row+1][col+1]);
                }
            }
        }

        return *min_element(matrix[0].begin(), matrix[0].end());
    }
};